### PR TITLE
massive rewrite to boto3

### PR DESCRIPTION
**Pull request by tedder**
##### ISSUE TYPE

- Feature Pull Request
##### COMPONENT NAME

cloudformation
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel 1f03801fcc) last updated 2016/09/29 14:29:35 (GMT -700)
  lib/ansible/modules/core: (cloudformation_to_boto3 5868eab238) last updated 2016/10/11 16:18:32 (GMT -700)
  lib/ansible/modules/extras: (devel 6c31d91fa5) last updated 2016/10/11 16:10:16 (GMT -700)
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

Completely rewrote cloudformation to boto3. I believe this is allowed/expected in 2.3.
- removed star-imports, which wasn't possible in Ansible 1.x
- boto doesn't have any of the modern features (most notably, changesets), so this rewrite goes all-in on boto3.
- tags are updateable, at least in boto3. Fix documentation.
- staying with "ansible yaml to json conversion" because I'm trying to keep this scoped properly. The next PR will have AWS-native yaml support.
- documented the output. Tried to leave it backwards-compatible but the changes to 'events' might break someone's flow. However, the existing data wasn't terribly useful so I don't assume it will hurt.
- split up the code into functions. This should make unit testing possible.
- added forward-facing code: 'six' for iterating, started using AWSRetry, common tag conversion.
- add todo list

I'm concerned that the listed maintainer, @jsmartin, is not around. He has very little activity on his github profile, and there are a bunch of pending pull requests for this module: #4517 #2819 #1224 #5112 #3293 #2314 #1460 #1370

### CODE REVIEW

**Comment by tedder** [comment]
cc @viper233

**Comment by jsmartin** [comment]
Looks like we got a new volunteer maintainer. :)  Feel free to own this module after your PR is merged.  I've no problems with it, but I am not following what the standard Ansible practice is for a switch over to boto3.  @defionscode?  We may want a a parallel CF module that is boto3 specific and deprecate this one as to not break anyone's existing playbooks.

**Comment by tedder** [comment]
I'll sign up to maintain but I don't want to be the only one. [Here's what I had found about the boto3/2.3 strategy](https://github.com/ansible/ansible-modules-extras/pull/1846#issuecomment-243834364). I thought about leaving the old module behind- if it's called _cloudformation, can it be accessed in plays as _cloudformation? Or should this one be boto3_cloudformation? I don't really like the idea of "new coke", as it should really be "old coke" and "coke".

**Comment by gregdek** [comment]
Thanks @tedder. To the current maintainers, @jsmartin please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by juliedavila** [comment]
cc @ryansb 

This looks good to me though I haven't tested. The only thing I'll say is that the `module_utils` imports have been moved to the top of the file which is different from the typical style for modules in ansible.

**Comment by tedder** [comment]
@defionscode I have the module_utils imports at line 200, just below the documentation. Do you mean they should be around line 20?

**Comment by juliedavila** [comment]
@tedder not quite, I mean more like this: https://github.com/ansible/ansible-modules-core/blob/devel/cloud/amazon/ec2_elb.py#L373-L378

**Comment by tedder** [comment]
That has the m_u imports at the bottom of the file, so I'm confused by what you mean. I can hop on IRC if it's easier.

**Comment by juliedavila** [comment]
right, in my original message I said `have been moved`, to mean that you moved them from where they were originally, to where they are now. I'm on irc, same nic

**Comment by tedder** [comment]
fixed.

File: cloud/amazon/cloudformation.py
Comment (ryansb) [inline cloud/amazon/cloudformation.py:17]: I'd recommend not committing a TODO list into the file, though all these would be great features to have :smile:

File: cloud/amazon/cloudformation.py
Comment (ryansb) [inline cloud/amazon/cloudformation.py:20]: Would you want to chat on IRC or expand in a "feature_idea" issue on how these would be used? Mostly I see changesets as an interactive thing where the operator can say "Wait wait, that's not what I meant" before applying.

File: cloud/amazon/cloudformation.py
Comment (ryansb) [inline cloud/amazon/cloudformation.py:70]: Clarification: this is a removal because it's included in the master AWS/EC2 doc fragment, the parameter still works identically.

File: cloud/amazon/cloudformation.py
Comment (ryansb) [inline cloud/amazon/cloudformation.py]: Instead of stringifying the traceback, use the `exception` kwarg to `fail_json`. It takes the formatted traceback (which doesn't require an arg inside an exception handler). 

```
except Exception as err:
    module.fail_json(msg=boto_exception(err), exception=traceback.format_exc())
```

**Review by ryansb — CHANGES_REQUESTED** [review]
I'll take this for a spin locally as well, but overall looks good. Glad to get the boto3 version of this out so early in the dev cycle.

**Comment by ryansb** [comment]
@tedder thanks for the quick turnaround, I can also add myself as a maintainer for this module since it's part of `core`.

File: cloud/amazon/cloudformation.py
Comment (tedder) [reply cloud/amazon/cloudformation.py:17]: I can remove them, but as soon as this is pulled I'll add those features. I'm trying to keep this one as the baseline before making further changes. Let me know.

File: cloud/amazon/cloudformation.py
Comment (tedder) [reply cloud/amazon/cloudformation.py:20]: I'm on ansible_dev to chat, but yeah, that's how I see it too. Honestly I want to prototype it a little and see how it looks- for instance, does it make sense to wait in the cloudformation module for the change approval, or to bail and finish on the next run?

File: cloud/amazon/cloudformation.py
Comment (tedder) [reply cloud/amazon/cloudformation.py:70]: Correct.

File: cloud/amazon/cloudformation.py
Comment (tedder) [reply cloud/amazon/cloudformation.py]: done!

File: cloud/amazon/cloudformation.py
Comment (ryansb) [inline cloud/amazon/cloudformation.py:257]: This loop is causing me a bit of a headache, when rerunning the task againsta a template (and parameter set) that have been deployed already it marks `changed=True`. Still trying to figure out why.

Comment (tedder) [reply cloud/amazon/cloudformation.py:257]: I'll investigate.

Comment (tedder) [reply cloud/amazon/cloudformation.py:257]: Oh right, it's because nothing happens and it returns the status of the LAST run. Time to investigate a little.

Comment (tedder) [reply cloud/amazon/cloudformation.py:257]: got it.

**Comment by gregdek** [comment]
Thanks @tedder for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by tedder** [comment]
Um, revisions done, ready_for_review.

**Comment by tedder** [comment]
Curse you, Shippable. It needs to be retriggered.

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
The fail on the rerun is something to do with one of the Ubuntu 16 packages in the test env, not due to this patch.

File: cloud/amazon/cloudformation.py
Comment (pauricthelodger) [inline cloud/amazon/cloudformation.py:363]: Don't suppose you could integrate #5066 which remove this yaml»json block entirely? Or could that PR be approved by one of the maintainers separately if ye want to maintain the scoping? =)

Comment (tedder) [reply cloud/amazon/cloudformation.py:363]: That's next on the list. I may rebase it or just rewrite it. Agree, it's the most obvious thing missing from this, but it was intentional, I wanted to reduce scope as much as possible, considering how major the changes are.

Comment (ryansb) [reply cloud/amazon/cloudformation.py:363]: Yeah, let's leave that for a separate PR. @pauricthelodger we haven't forgotten about supporting CFN's new native YAML.

**Comment by ryansb** [comment]
@tedder I made some modifications to this PR to clear up documentation and some return values (specifically returning exceptions in more places). If it looks good to you, I can ship it out. I'm in #ansible-devel if you want to discuss as well.

**Comment by tedder** [comment]
Closing, I like Ryan's changes :)

File: cloud/amazon/cloudformation.py
Comment (pauricthelodger) [reply cloud/amazon/cloudformation.py:363]: Sound, cheers folks and congrats on the successful merge =)

**Pull request closed without merging**
